Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/CodeBuilders.kt

```diff
@@ -34,42 +34,81 @@ inline fun buildNativeCodeLines(scope: NativeScope, block: NativeCodeBuilder.()
     return builder.lines
 }
 
+private class Block(val nesting: Int, val start: String, val end: String) {
+    val prologue = mutableListOf<String>()
+    val body = mutableListOf<String>()
+    val epilogue = mutableListOf<String>()
+
+    fun indent(line: String) = "    ".repeat(nesting) + line
+    fun indentBraces(line: String) = "    ".repeat(nesting - 1) + line
+}
+
 class KotlinCodeBuilder(val scope: KotlinScope) {
-    private val lines = mutableListOf<String>()
 
-    private val freeStack = mutableListOf<String>()
-    private val nesting get() = freeStack.size
+    private val blocks = mutableListOf<Block>()
+
+    init {
+        pushBlock("", "")
+    }
 
     fun out(line: String) {
-        lines.add("    ".repeat(nesting) + line)
+        currentBlock().body += line
     }
 
     private var memScoped = false
-    fun pushMemScoped() {
+    private var clearIndex = -1
+    fun pushMemScoped(explicitClear: Boolean = true) {
+        if (!explicitClear) {
+            if (clearIndex != -1) {
+                blocks[0].epilogue[clearIndex] = "/* memScope.clear() is not needed here. */"
+            }
+        }
+
         if (!memScoped) {
             memScoped = true
-            pushBlock("memScoped {")
+            blocks[0].prologue += "val memScope = Arena()"
+            if (explicitClear) {
+                clearIndex = blocks[0].epilogue.size
+                blocks[0].epilogue += "memScope.clear()"
+            }
         }
     }
 
-    fun pushBlock(line: String, free: String = "") {
-        out(line)
-        freeStack.add(free)
+    fun getNativePointer(name: String): String {
+        return "$name?.getPointer(memScope)"
     }
 
-    private fun popBlocks() {
-        while (freeStack.isNotEmpty()) {
-            val free = freeStack.last()
-            freeStack.removeAt(freeStack.lastIndex)
-            out("} $free".trim())
+    fun returnResult(result: String, returnsVoid: Boolean) {
+        if (returnsVoid) {
+            currentBlock().body += result
+        } else {
+            currentBlock().body += "val _r_ = $result"
+            currentBlock().epilogue += "return _r_"
```

Comment:
The entire refactoring doesn't look correct to me.
For example, outer blocks' epilogues aren't going to be executed after such a return.

Reply:
Reverted back to memScoped, while keeping more powerful code structure generator.